dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/EventDataValuesValidationHook.java

```diff
@@ -108,17 +110,34 @@ private void validateDataElement( ValidationErrorReporter reporter, DataElement
         }
         else
         {
+            validateCompulsoryDataElement( reporter, dataElement, programStage, dataValue );
             validateFileNotAlreadyAssigned( reporter, dataValue, dataElement );
         }
     }
 
-    private void validateDataValueDataElementIsConnectedToProgramStage( ValidationErrorReporter reporter,
-        TrackerImportValidationContext ctx, Event event )
+    private void validateCompulsoryDataElement( ValidationErrorReporter reporter, DataElement dataElement,
+        ProgramStage programStage, DataValue dataValue )
     {
-        if ( StringUtils.isEmpty( event.getProgramStage() ) )
-            return;
+        Optional<ProgramStageDataElement> optionalPsde = Optional.ofNullable( programStage )
+            .map( ps -> ps.getProgramStageDataElements().stream() ).flatMap( psdes -> psdes
+                .filter(
+                    psde -> psde.getDataElement().getUid().equals( dataElement.getUid() ) && psde.isCompulsory() )
+                .findFirst() );
+
+        if ( optionalPsde.isPresent() && dataValue.getValue() == null )
+        {
+            addError( reporter, E1076, DataElement.class.getSimpleName(),
+                dataElement.getUid() );
+        }
 
-        ProgramStage programStage = ctx.getProgramStage( event.getProgramStage() );
+    }
+
+    private void validateDataValueDataElementIsConnectedToProgramStage( ValidationErrorReporter reporter, Event event,
+        ProgramStage programStage )
+    {
+
+        if ( null == programStage )
```

Comment:
ProgramStage must not be null here.
We could add checkNotNull when getting from the Preheat and remove this check here